/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

 //iterative approach
class Solution {
    public ListNode reverseList(ListNode head) {
      ListNode prev=null;
      ListNode temp=head;
      while(temp!=null){
        ListNode front=temp.next;
        temp.next=prev;
        
        prev=temp;
        temp=front;
      }
      return prev;
    }
}


 //recursive approach
class Solution {
    public ListNode reverseList(ListNode head) {
       if(head==null || head.next==null) return head;
       
        ListNode newHead=reverseList(head.next);
        ListNode front=head.next;
        head.next=null;
        front.next=head;
        return newHead;
    }
}


